### PR TITLE
Added Json API Relationship Tests (self links)

### PR DESCRIPTION
**Pull request by DellanX**
# Description

Added Test Cases for the JsonAPI Entity Store ToMany Relationships Feature discussed in #7 
Seeing that 2 test conditions fail, more development may be needed on this feature.
* FindAll - does not retrieve Fields
* Include - does not attach related object

## Type of change

Test Coverage
# How Has This Been Tested?

Tests were validated using a Red-Green approach. (Fault Insertion)
Each test case was added using invalid expectations. Once the Failures were detected, the test was updated with the valid expecations.

For the 2 tests that are still failing, the lines in question were forced to pass instead of having a fault insertion, to ensure that the logic was operating correctly.

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by DellanX** [comment]
Okay, I think with commit 0a48cdc, I was able to fix both failing tests; however, it also looks like there is an issue with your backend when it comes to querying include statements on relationships.

Something is up with SQLite and the JsonApi Server .net library. 
```json
{
  "links": {
    "self": "http://localhost:20001/users/2088/messages?include=user",
    "first": "http://localhost:20001/users/2088/messages?include=user",
    "last": "http://localhost:20001/users/2088/messages?include=user"
  },
  "errors": [
    {
      "id": "4ae5fab0-6005-4370-9a5e-418ea5affa8e",
      "status": "500",
      "title": "An unhandled error occurred while processing this request.",
      "detail": "Translating this query requires the SQL APPLY operation, which is not supported on SQLite."
    }
  ],
  "included": [
    
  ],
  "meta": {
    "total": 3
  }
}
```

The query generated looks correct, and after copying the query over to my personal application, which runs a MySQL database, the server was able to make the query successfully. Further, it populated the fields as expected.

Another option may be to add better configuration to the selfLink parameter I added to your code.
You could maybe query the self link, giving the identifier objects, then take the set of data to run a secondary query to gather the fields.

**Comment by dpimonov** [comment]
Hey, thanks for the PR. I like how you implemented tests with separate setup function. With self link parameter - I need to dig a little bit the topic. The spec is unclear here - you are correct. Also I don't like passing bool properties as parameters. Maybe creating a separate method will be better. But it is a minor thing for which I don't have concrete opinion right now.

So far I have nothing to correct. Then if you have time - implement please tests for to-one relation similar to what you already did and also doc section for these two.

```
Something is up with SQLite and the JsonApi Server .net library.
```

That's sad - query looks correct. I want to avoid database setup stuff for tests and two queries of course. I will play with it also when have time.  Thanks for pointing this out.

**Comment by DellanX** [comment]
> Also I don't like passing bool properties as parameters. Maybe creating a separate method will be better. But it is a minor thing for which I don't have concrete opinion right now.

Me either, though I didn't necessarily know what was better with the direction of your code.
I wouldn't mind duplicating the method, but would need to know what naming conventions you'd want to use for this scenario.

> So far I have nothing to correct. Then if you have time - implement please tests for to-one relation similar to what you already did and also doc section for these two.

Okay, I can start on that.

> That's sad - query looks correct. I want to avoid database setup stuff for tests and two queries of course. I will play with it also when have time. Thanks for pointing this out.

May need to either translate the APPLY statements via .net somewhere, if you can intercept the DB calls.
Or switch to a different database system... Either will probably take a lot of time.

**Comment by DellanX** [comment]
Okay, with my last commit, I added tests and features.
If you have a chance, do tell me if this is going in the direction you want, or if you want me to change anything up.

Otherwise I'll continue working on the include test coverage later.

# Added Tests
I've added support for toOne testing, which has the following features:
* Fetching toOne relationship
* Updating toOne relationship
* Removing toOne relationship
* Fetching with Include Clause toOne relationship

# Added Features
I also added support to the toOne provider for querying the relationship; however, this was a bit hacky, as `include` requires the BulkQuery command support.

I ended up wrapping the individual entity data into a temporary pagination collection.

# Notes
After adding test support for toOne, I realize that I may need to rework my `include` tests to cover both `include` and `includeCollection`

I'll be looking into this tomorrow.

**Comment by dpimonov** [comment]
Looks nice! About include commands - i would suggest to ignore it for to-one relations for now. It requires a little bit different approach as we cannot simply return browse command for a single relation. The browse command provides many things which are not compatible with it. First `EntityStore` package should be reorganized a bit to somehow seperate include command from browse commands - then it can be properly implemented. For now the main porpose it to simply add / update / remove to-one relations.

**Comment by dpimonov** [comment]
```
I wouldn't mind duplicating the method, but would need to know what naming conventions you'd want to use for this scenario.
```

Well as this logic with `false` flag is used only in `createRelationshipBrowseLinkObject` - i think the better option will be simply put this part directly there and remove bool flag completely.

```
public createRelationshipBrowseLinkObject(....)
{
    const resourceIdentifierLinkObject = this.createResourceIdentifierLinkObject(typeMetadata, entityOrIdentifier);
    const relationship = propertyMetadata.serializedPropertyName;
    const relationshipLinkObject = `${resourceIdentifierLinkObject}/${relationship}`;
    const linkObject = this.createBrowseLinkObject(relationshipLinkObject, browseCommand);

    return linkObject;
}
```

**Comment by DellanX** [comment]
Okay, I'll make the changes.

As for fetch support, I figure I can re-implement that functionality using the `find` function and the `executeQueryCommand` instead of `findOne` and `executeBulkQueryCommand`. Unless that's also not recommended.

At some point, I figure it'd be nice to add relationship support to the `EntityStore` package to simplify implementation for other entity providers. I assume that there aren't any plans for this yet?

Perhaps what I can do is take a look into that after this PR.
(Though I may attempt an M-M relationship provider first, as that should be relatively straightforward at this point of time)

**Comment by dpimonov** [comment]
```
As for fetch support, I figure I can re-implement that functionality using the find function and the executeQueryCommand instead of findOne and executeBulkQueryCommand. Unless that's also not recommended.
```

Good finding! This is more correct way. Lets keep it.

```
At some point, I figure it'd be nice to add relationship support to the EntityStore package to simplify implementation for other entity providers. I assume that there aren't any plans for this yet?
```

I am not sure if it makes sense for other providers except JsonApi. Normally one can use filtering API to filter relations and save entities normally without using this funky identifier objects. Besides - making it part of `EntityStore` will force all providers to implement it. If there will be one more use case in the future in other provider - then we can think about moving it to the main package. Right now it is not clear if this change worth efforts. Besides there are some plans for lazy api. Which mean one can load for example `User` and later simply access `Messages` property and on this access a query will be sent automatically. But right now it is not a high priority feature and requires a bit of change detection.

About plans in general:

1) Right now in my free time I am busy with declarative api and propper plugin integration api. Maybe you have already noticed that `TypeManager` provides a way to configure everything without decorators and also ability to have multiple `TypeManager` instances. There is currently no propper way of using this set of functions in `EntityStore` and `JsonApiEntityProvider`. They are coupled to decorators and static instances. This is the first thing on the road to be improved in both packages. I am working on it right now.

2) `JSON:API` is now in v1.1 but so far I don't know what was added. Maybe one or another improvement might occur. Also right now there is only support for `JSON.NET` filter and paginate visitors in this library. There are other server libraries available which worth implementing visitors so we can attract more developers. Besides `EntityStore` might also be used for implementation a server side of `JSON:API` but for now I want to cover only clients as server side is a bit more complex.

3) Implementing more entity providers is a next big topic. Main step is to cover all possible database providers with lets call it `DatabaseEntityProvider` or `SqlEntityProvider` plugin as a second implementation. However this requires specification first and blocked by point 1. In comparison to `JsonApiEntityProvider` query and deserialization part is relatively easy but configuration with table, column, types and etc should be well thought. And there are many other topics besides that with this plugin.

4) [ODATA](https://www.odata.org/) spec looks interesting. It is almost the same idea as `JSON:API` but a way more advanced and might required implementing method metadata in `TypeManager`. However this topic also looks really interesting.

These are main plans. If you want to help further - it is a good way to go. However these topics should be specified first and right now more about discussion, sketches and investigation what already available in the market and how to make it super simple and efficient. 😊

**Comment by DellanX** [comment]
# Moving this to Entity Framework
TL;DR; Definitely not a part of the Pull Request.

> am not sure if it makes sense for other providers except JsonApi. 

As you said, it probably wouldn't be easy to implement an abstract system for this. 
If I investigated into this, it'd be from the mindset of creating a base relationship provider in EntityStore, which translates a relationship call into the general entitySet equivalent command.

Then, if an entityProvider (such as JsonAPI) has a better way to handle relationships, it may override the relationship behavior.

Again, as you said, it's probably not something that can be squeezed in. If I explore it further, I'll come up with a demo and create a draft PR for it on the EntityStore.TS repo

# Your plans
1. I did not know that was possible, while for me the decorator feature is my favorite thing about this package, I do figure there is someone out there who cannot use it. It'll be nice there. 
2. JSON:API 1.1, they added the atomic operation extension: [atomic operations](https://jsonapi.org/ext/atomic/)
3. I do agree, more providers would be nice. If ever I decided to migrate my server from PHP to something like TS, then the MySql provider would be a must.

# This PR
Is there anything else you need from me to close out this PR?

**Comment by dpimonov** [comment]
Thanks for the feedback! About this PR - documentation part (README) of how to work with relations is missing. We need a new section for it.

**Comment by DellanX** [comment]
I added some documentation to the readme as a rough draft.
Do note, documentation is by far my weakest skill.

**Comment by dpimonov** [comment]
Thanks! You did a great job!

```
Do note, documentation is by far my weakest skill.
```

For me it looks really nice! I also do not spent much time on describing this plugin so far. 😁

As you opened another issue: should we include docker stuff with this PR or would you like to have a separate one?

**Comment by DellanX** [comment]
I think the docker stuff should be another PR.

It'll require recreating the backend, which is a lot of work unrelated to the relationship functionality.

**Comment by dpimonov** [comment]
```
It'll require recreating the backend, which is a lot of work unrelated to the relationship functionality.
```

Hmm. It is only one click in the VS. I will add it and then merge your changes. 😄

**Comment by dpimonov** [comment]
I've pushed changes to the main branch. You can check if it works for you. I used postgres as a replacement as I am more familiar with it. After pulling changes VS should automatically detect docker environment and replace start button with docker compose. It it does not work for some reason then set it as a startup project manually.

image

Tell me if your tests pass after this replacement. If not - you are free to exchange postgres with MySQL. 😄
